File: backend/sdk/typescript/sbh-sdk.ts
/**
 * SBH SDK for TypeScript (React/Next.js)
 * Provides auth helpers, RBAC hooks, API client, and event tracking.
 */

import React, { createContext, useContext, useState, useEffect, ReactNode } from 'react';
import axios, { AxiosInstance, AxiosResponse } from 'axios';

// Types
export interface SBHUser {
  id: string;
  email: string;
  name: string;
  role: string;
  tenantId: string;
  permissions: string[];
}

export interface SBHContext {
  user: SBHUser | null;
  tenantId: string | null;
  isAuthenticated: boolean;
  isLoading: boolean;
}

export interface SBHConfig {
  apiUrl: string;
  authToken?: string;
}

export interface SBHApiResponse<T = any> {
  data: T;
  meta?: any;
  errors?: any[];
}

// API Client
export class SBHApiClient {
  private client: AxiosInstance;
  private config: SBHConfig;

  constructor(config: SBHConfig) {
    this.config = config;
    this.client = axios.create({
      baseURL: config.apiUrl,
      timeout: 30000,
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (config.authToken) {
      this.client.defaults.headers.common['Authorization'] = `Bearer ${config.authToken}`;
    }

    // Add response interceptor for error handling
    this.client.interceptors.response.use(
      (response) => response,
      (error) => {
        if (error.response?.status === 401) {
          console.error('Authentication failed');
        } else if (error.response?.status === 403) {
          console.error('Access denied');
        }
        throw error;
      }
    );
  }

  async get<T>(endpoint: string, params?: any): Promise<SBHApiResponse<T>> {
    const response: AxiosResponse<SBHApiResponse<T>> = await this.client.get(endpoint, { params });
    return response.data;
  }

  async post<T>(endpoint: string, data?: any): Promise<SBHApiResponse<T>> {
    const response: AxiosResponse<SBHApiResponse<T>> = await this.client.post(endpoint, data);
    return response.data;
  }

  async put<T>(endpoint: string, data?: any): Promise<SBHApiResponse<T>> {
    const response: AxiosResponse<SBHApiResponse<T>> = await this.client.put(endpoint, data);
    return response.data;
  }

  async delete<T>(endpoint: string): Promise<SBHApiResponse<T>> {
    const response: AxiosResponse<SBHApiResponse<T>> = await this.client.delete(endpoint);
    return response.data;
  }

  // Auth methods
  async login(credentials: { email: string; password: string }): Promise<{ token: string; user: SBHUser }> {
    const response = await this.post('/api/auth/login', credentials);
    return response.data;
  }

  async getProfile(): Promise<SBHUser> {
    const response = await this.get('/api/auth/profile');
    return response.data;
  }

  async logout(): Promise<void> {
    await this.post('/api/auth/logout');
  }

  // Analytics methods
  async track(event: string, properties?: any, userId?: string): Promise<void> {
    await this.post('/api/analytics/track', {
      event,
      properties: properties || {},
      userId,
      timestamp: new Date().toISOString(),
    });
  }

  async identify(userId: string, traits?: any): Promise<void> {
    await this.post('/api/analytics/identify', {
      userId,
      traits: traits || {},
      timestamp: new Date().toISOString(),
    });
  }
}

// React Context
const SBHContext = createContext<{
  api: SBHApiClient;
  context: SBHContext;
  login: (credentials: { email: string; password: string }) => Promise<void>;
  logout: () => Promise<void>;
  track: (event: string, properties?: any) => Promise<void>;
} | null>(null);

// Provider Component
interface SBHProviderProps {
  children: ReactNode;
  config: SBHConfig;
}

export function SBHProvider({ children, config }: SBHProviderProps) {
  const [api] = useState(() => new SBHApiClient(config));
  const [context, setContext] = useState<SBHContext>({
    user: null,
    tenantId: null,
    isAuthenticated: false,
    isLoading: true,
  });

  useEffect(() => {
    // Check for existing auth token
    const token = localStorage.getItem('sbh_auth_token');
    if (token) {
      api.config.authToken = token;
      loadProfile();
    } else {
      setContext(prev => ({ ...prev, isLoading: false }));
    }
  }, []);

  const loadProfile = async () => {
    try {
      const user = await api.getProfile();
      setContext({
        user,
        tenantId: user.tenantId,
        isAuthenticated: true,
        isLoading: false,
      });
    } catch (error) {
      localStorage.removeItem('sbh_auth_token');
      setContext({
        user: null,
        tenantId: null,
        isAuthenticated: false,
        isLoading: false,
      });
    }
  };

  const login = async (credentials: { email: string; password: string }) => {
    try {
      const { token, user } = await api.login(credentials);
      localStorage.setItem('sbh_auth_token', token);
      api.config.authToken = token;
      
      setContext({
        user,
        tenantId: user.tenantId,
        isAuthenticated: true,
        isLoading: false,
      });
    } catch (error) {
      throw error;
    }
  };

  const logout = async () => {
    try {
      await api.logout();
    } catch (error) {
      console.error('Logout error:', error);
    } finally {
      localStorage.removeItem('sbh_auth_token');
      api.config.authToken = undefined;
      setContext({
        user: null,
        tenantId: null,
        isAuthenticated: false,
        isLoading: false,
      });
    }
  };

  const track = async (event: string, properties?: any) => {
    try {
      await api.track(event, properties, context.user?.id);
    } catch (error) {
      console.error('Analytics error:', error);
    }
  };

  return (
    <SBHContext.Provider value={{ api, context, login, logout, track }}>
      {children}
    </SBHContext.Provider>
  );
}

// Hook
export function useSBH() {
  const context = useContext(SBHContext);
  if (!context) {
    throw new Error('useSBH must be used within an SBHProvider');
  }
  return context;
}

// API Hook
export function useSBHApi() {
  const { api } = useSBH();
  return api;
}

// Auth Hook
export function useSBHAuth() {
  const { context, login, logout } = useSBH();
  return {
    user: context.user,
    isAuthenticated: context.isAuthenticated,
    isLoading: context.isLoading,
    login,
    logout,
  };
}

// Analytics Hook
export function useSBHAnalytics() {
  const { track } = useSBH();
  return { track };
}

// RBAC Hooks
export function useRequireAuth() {
  const { context } = useSBH();
  
  if (context.isLoading) {
    return { loading: true };
  }
  
  if (!context.isAuthenticated) {
    throw new Error('Authentication required');
  }
  
  return { loading: false };
}

export function useRequireRole(roles: string[]) {
  const { context } = useSBH();
  
  if (context.isLoading) {
    return { loading: true };
  }
  
  if (!context.isAuthenticated) {
    throw new Error('Authentication required');
  }
  
  if (!context.user || !roles.includes(context.user.role)) {
    throw new Error(`Role required: ${roles.join(', ')}`);
  }
  
  return { loading: false };
}

export function useRequirePermission(permission: string) {
  const { context } = useSBH();
  
  if (context.isLoading) {
    return { loading: true };
  }
  
  if (!context.isAuthenticated) {
    throw new Error('Authentication required');
  }
  
  if (!context.user || !context.user.permissions.includes(permission)) {
    throw new Error(`Permission required: ${permission}`);
  }
  
  return { loading: false };
}

// API Hooks
export function useApi<T = any>(endpoint: string, params?: any) {
  const { api } = useSBH();
  const [data, setData] = useState<T | null>(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<Error | null>(null);

  useEffect(() => {
    const fetchData = async () => {
      try {
        setLoading(true);
        const response = await api.get<T>(endpoint, params);
        setData(response.data);
        setError(null);
      } catch (err) {
        setError(err as Error);
      } finally {
        setLoading(false);
      }
    };

    fetchData();
  }, [endpoint, JSON.stringify(params)]);

  return { data, loading, error };
}

export function useApiMutation<T = any>(endpoint: string) {
  const { api } = useSBH();
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<Error | null>(null);

  const mutate = async (data?: any): Promise<T | null> => {
    try {
      setLoading(true);
      setError(null);
      const response = await api.post<T>(endpoint, data);
      return response.data;
    } catch (err) {
      setError(err as Error);
      return null;
    } finally {
      setLoading(false);
    }
  };

  return { mutate, loading, error };
}

// Example Components
export function ContactsList() {
  const { data, loading, error } = useApi<any[]>('/api/contacts');
  const { track } = useSBHAnalytics();

  useEffect(() => {
    track('contacts.list.viewed');
  }, []);

  if (loading) return <div>Loading contacts...</div>;
  if (error) return <div>Error loading contacts: {error.message}</div>;

  return (
    <ul>
      {data?.map(contact => (
        <li key={contact.id}>{contact.first_name} {contact.last_name}</li>
      ))}
    </ul>
  );
}

export function CreateContactForm() {
  const { mutate, loading } = useApiMutation('/api/contacts');
  const { track } = useSBHAnalytics();

  const handleSubmit = async (formData: any) => {
    const result = await mutate(formData);
    if (result) {
      track('contact.created', { contactId: result.id });
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      {/* Form fields */}
      <button type="submit" disabled={loading}>
        {loading ? 'Creating...' : 'Create Contact'}
      </button>
    </form>
  );
}

// Protected Component Example
export function ProtectedContactsList() {
  useRequirePermission('contacts.read');

  return <ContactsList />;
}

export function AdminOnlyComponent() {
  useRequireRole(['admin', 'owner']);

  return <div>Admin only content</div>;
}

File: backend/sdk/python/sbh_sdk.py
"""
SBH SDK for Python (FastAPI/Flask)
Provides auth helpers, RBAC decorators, API client, and event tracking.
"""

import functools
import json
import logging
from typing import Dict, Any, Optional, Callable, List
from dataclasses import dataclass
from datetime import datetime
import requests

logger = logging.getLogger(__name__)


@dataclass
class SBHUser:
    """SBH user information."""
    id: str
    email: str
    name: str
    role: str
    tenant_id: str
    permissions: List[str]


@dataclass
class SBHContext:
    """SBH request context."""
    user: SBHUser
    tenant_id: str
    request_id: str
    timestamp: datetime


class SBHAuth:
    """SBH Authentication helper."""
    
    def __init__(self, api_url: str = "http://localhost:5001"):
        self.api_url = api_url
        self._current_user: Optional[SBHUser] = None
        self._current_tenant: Optional[str] = None
    
    def get_current_user(self) -> Optional[SBHUser]:
        """Get current authenticated user."""
        return self._current_user
    
    def get_current_tenant(self) -> Optional[str]:
        """Get current tenant ID."""
        return self._current_tenant or (self._current_user.tenant_id if self._current_user else None)
    
    def set_user(self, user: SBHUser):
        """Set current user."""
        self._current_user = user
        self._current_tenant = user.tenant_id
    
    def require_auth(self, f: Callable) -> Callable:
        """Decorator to require authentication."""
        @functools.wraps(f)
        def wrapper(*args, **kwargs):
            if not self._current_user:
                raise ValueError("Authentication required")
            return f(*args, **kwargs)
        return wrapper
    
    def require_role(self, roles: List[str]) -> Callable:
        """Decorator to require specific roles."""
        def decorator(f: Callable) -> Callable:
            @functools.wraps(f)
            def wrapper(*args, **kwargs):
                if not self._current_user:
                    raise ValueError("Authentication required")
                if self._current_user.role not in roles:
                    raise ValueError(f"Role required: {roles}")
                return f(*args, **kwargs)
            return wrapper
        return decorator
    
    def require_permission(self, permission: str) -> Callable:
        """Decorator to require specific permission."""
        def decorator(f: Callable) -> Callable:
            @functools.wraps(f)
            def wrapper(*args, **kwargs):
                if not self._current_user:
                    raise ValueError("Authentication required")
                if permission not in self._current_user.permissions:
                    raise ValueError(f"Permission required: {permission}")
                return f(*args, **kwargs)
            return wrapper
        return decorator


class SBHDatabase:
    """SBH Database helper."""
    
    def __init__(self, db_url: str):
        self.db_url = db_url
    
    def query(self, sql: str, params: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:
        """Execute a SQL query with tenant isolation."""
        # This is a simplified implementation
        # In a real implementation, you would use SQLAlchemy or similar
        if params and 'tid' in params:
            # Ensure tenant isolation
            sql = f"{sql} AND tenant_id = :tid"
        
        logger.info(f"Executing query: {sql}")
        # Placeholder for actual database query
        return []
    
    def execute(self, sql: str, params: Optional[Dict[str, Any]] = None) -> int:
        """Execute a SQL statement."""
        logger.info(f"Executing statement: {sql}")
        # Placeholder for actual database execution
        return 1
    
    def get_by_id(self, table: str, id: str, tenant_id: str) -> Optional[Dict[str, Any]]:
        """Get record by ID with tenant isolation."""
        sql = f"SELECT * FROM {table} WHERE id = :id AND tenant_id = :tid"
        result = self.query(sql, {"id": id, "tid": tenant_id})
        return result[0] if result else None
    
    def list(self, table: str, tenant_id: str, filters: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:
        """List records with tenant isolation."""
        sql = f"SELECT * FROM {table} WHERE tenant_id = :tid"
        params = {"tid": tenant_id}
        
        if filters:
            for key, value in filters.items():
                sql += f" AND {key} = :{key}"
                params[key] = value
        
        return self.query(sql, params)
    
    def create(self, table: str, data: Dict[str, Any], tenant_id: str) -> str:
        """Create a new record with tenant isolation."""
        data['tenant_id'] = tenant_id
        data['created_at'] = datetime.utcnow().isoformat()
        data['updated_at'] = datetime.utcnow().isoformat()
        
        columns = ', '.join(data.keys())
        placeholders = ', '.join([f':{k}' for k in data.keys()])
        sql = f"INSERT INTO {table} ({columns}) VALUES ({placeholders}) RETURNING id"
        
        # Placeholder for actual insert
        return "generated-id"
    
    def update(self, table: str, id: str, data: Dict[str, Any], tenant_id: str) -> bool:
        """Update a record with tenant isolation."""
        data['updated_at'] = datetime.utcnow().isoformat()
        
        set_clause = ', '.join([f"{k} = :{k}" for k in data.keys()])
        sql = f"UPDATE {table} SET {set_clause} WHERE id = :id AND tenant_id = :tid"
        
        data['id'] = id
        data['tid'] = tenant_id
        
        return self.execute(sql, data) > 0
    
    def delete(self, table: str, id: str, tenant_id: str) -> bool:
        """Delete a record with tenant isolation."""
        sql = f"DELETE FROM {table} WHERE id = :id AND tenant_id = :tid"
        return self.execute(sql, {"id": id, "tid": tenant_id}) > 0


class SBHFiles:
    """SBH File storage helper."""
    
    def __init__(self, storage_url: str = "http://localhost:5001"):
        self.storage_url = storage_url
    
    def upload(self, file_path: str, tenant_id: str, folder: str = "uploads") -> str:
        """Upload a file to SBH storage."""
        # Placeholder for actual file upload
        logger.info(f"Uploading file {file_path} to {folder} for tenant {tenant_id}")
        return f"https://storage.sbh.com/{tenant_id}/{folder}/file-id"
    
    def download(self, file_url: str) -> bytes:
        """Download a file from SBH storage."""
        # Placeholder for actual file download
        logger.info(f"Downloading file from {file_url}")
        return b"file content"
    
    def delete(self, file_url: str) -> bool:
        """Delete a file from SBH storage."""
        # Placeholder for actual file deletion
        logger.info(f"Deleting file {file_url}")
        return True


class SBHAnalytics:
    """SBH Analytics helper."""
    
    def __init__(self, api_url: str = "http://localhost:5001"):
        self.api_url = api_url
    
    def track(self, event: str, properties: Optional[Dict[str, Any]] = None, user_id: Optional[str] = None):
        """Track an analytics event."""
        data = {
            "event": event,
            "properties": properties or {},
            "timestamp": datetime.utcnow().isoformat(),
            "user_id": user_id
        }
        
        try:
            response = requests.post(f"{self.api_url}/api/analytics/track", json=data)
            response.raise_for_status()
            logger.info(f"Tracked event: {event}")
        except Exception as e:
            logger.error(f"Failed to track event {event}: {e}")
    
    def identify(self, user_id: str, traits: Optional[Dict[str, Any]] = None):
        """Identify a user for analytics."""
        data = {
            "user_id": user_id,
            "traits": traits or {},
            "timestamp": datetime.utcnow().isoformat()
        }
        
        try:
            response = requests.post(f"{self.api_url}/api/analytics/identify", json=data)
            response.raise_for_status()
            logger.info(f"Identified user: {user_id}")
        except Exception as e:
            logger.error(f"Failed to identify user {user_id}: {e}")


class SBHAPI:
    """SBH API client."""
    
    def __init__(self, api_url: str = "http://localhost:5001", auth_token: Optional[str] = None):
        self.api_url = api_url
        self.auth_token = auth_token
        self.session = requests.Session()
        
        if auth_token:
            self.session.headers.update({"Authorization": f"Bearer {auth_token}"})
    
    def get(self, endpoint: str, params: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Make a GET request to SBH API."""
        url = f"{self.api_url}{endpoint}"
        response = self.session.get(url, params=params)
        response.raise_for_status()
        return response.json()
    
    def post(self, endpoint: str, data: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Make a POST request to SBH API."""
        url = f"{self.api_url}{endpoint}"
        response = self.session.post(url, json=data)
        response.raise_for_status()
        return response.json()
    
    def put(self, endpoint: str, data: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Make a PUT request to SBH API."""
        url = f"{self.api_url}{endpoint}"
        response = self.session.put(url, json=data)
        response.raise_for_status()
        return response.json()
    
    def delete(self, endpoint: str) -> Dict[str, Any]:
        """Make a DELETE request to SBH API."""
        url = f"{self.api_url}{endpoint}"
        response = self.session.delete(url)
        response.raise_for_status()
        return response.json()


# Global instances
auth = SBHAuth()
db = SBHDatabase("postgresql://user:pass@localhost/sbh")
files = SBHFiles()
analytics = SBHAnalytics()
api = SBHAPI()


# Convenience decorators
def require_auth(f: Callable) -> Callable:
    """Require authentication decorator."""
    return auth.require_auth(f)


def require_role(roles: List[str]) -> Callable:
    """Require role decorator."""
    return auth.require_role(roles)


def require_permission(permission: str) -> Callable:
    """Require permission decorator."""
    return auth.require_permission(permission)


# Example usage:
@require_auth
@require_role(["admin", "member"])
def get_contacts():
    """Get contacts for current tenant."""
    tenant_id = auth.get_current_tenant()
    return db.list("contacts", tenant_id)


@require_auth
@require_permission("contacts.create")
def create_contact(contact_data: Dict[str, Any]):
    """Create a new contact."""
    tenant_id = auth.get_current_tenant()
    contact_id = db.create("contacts", contact_data, tenant_id)
    
    # Track the event
    analytics.track("contact.created", {
        "contact_id": contact_id,
        "tenant_id": tenant_id
    }, auth.get_current_user().id)
    
    return contact_id
